### PR TITLE
Scheduler

### LINKED ISSUE #589: Schedulers
During training, some hyper-parameters (e.g learning rate, entropy target, etc.) need to change with iterations. It's tedious to separately implementing the scheduling for each different hyper-parameter. We can use the method in this post to implement schedulers which can be used in the same way as an int or float: https://stackoverflow.com/questions/1117149/does-python-have-conversion-operators

### PR DESCRIPTION
**Pull request by emailweixu**
Support target entropy scheduling and learning rate scheduling

### CODE REVIEW

File: alf/utils/schedulers.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+class LinearScheduler(Scheduler):
+    """The value is linearly changed in each defined region of progress."""
+
+    def __init__(self, progress_type, schedule):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+            schedule (list[tuple]): each tuple is a pair of (progress, value)
+                which means that if current progress is less than progress,
+                the current value will be value.
```
Comment (Haichao-Zhang) [inline alf/utils/schedulers.py]: the current value -> a linearly interpolated value

File: alf/trainers/policy_trainer.py [hunk 2]
```diff
@@ -230,8 +238,9 @@ def train(self):
                 ps.print_callees()
 
                 logging.info(s.getvalue())
-        finally:
```
Comment (hnyu) [inline alf/trainers/policy_trainer.py:233]: Why making this change? Currently it will save the checkpoint if an error occurs, which is good.

File: alf/utils/schedulers.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        super().__init__(progress_type)
+        self._progresses, self._values = zip(*schedule)
+        self._index = 0
+
+    def __call__(self):
+        progress = self.progress()
+        index = self._index
+        progresses = self._progresses
+        if progress < progresses[index]:
+            while index > 0 and progress < progresses[index - 1]:
```
Comment (hnyu) [inline alf/utils/schedulers.py]: In what situation we need to backstep the index? It seems that the progress is always incrementing

File: alf/utils/schedulers.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+@gin.configurable
+class LinearScheduler(Scheduler):
+    """The value is linearly changed in each defined region of progress."""
+
+    def __init__(self, progress_type, schedule):
+        """
+        Args:
+            progress_type (str): one of "percent", "iterations", "env_steps"
+            schedule (list[tuple]): each tuple is a pair of (progress, value)
```
Comment (hnyu) [inline alf/utils/schedulers.py:95]: This comment is still for stepscheduler.

**Review by hnyu — CHANGES_REQUESTED** [review]

**Review by Haichao-Zhang — DISMISSED** [review]

File: alf/utils/schedulers.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                which means that if the current progress between progress[i-1]
+                and progress[i], a linear interpolation between value[i-1] and
+                value[i] will be used. progress[0] must be 0.
```
Comment (hnyu) [inline alf/utils/schedulers.py]: Maybe also add a sentence about if the current progress is greater than progress[-1]? (then it's always the same last value)

**Review by hnyu — APPROVED** [review]
